var/spack/repos/builtin/packages/zlib-ng/package.py

```diff
@@ -69,9 +69,13 @@ class ZlibNg(AutotoolsPackage, CMakePackage):
     @property
     def libs(self):
         name = "libz" if self.spec.satisfies("+compat") else "libz-ng"
-        return find_libraries(
-            name, root=self.prefix, recursive=True, shared=self.spec.satisfies("+shared")
-        )
+        shared = self.spec.satisfies("+shared")
+        libs = find_libraries(name, root=self.prefix, recursive=True, shared=shared)
+        # With some compilers, e.g. cce@17, building a shared library may fail,
+        # so the build system just builds a static version.
```

Comment:
The issue here is that `cce@17` uses the LLD linker which fails the autodetection of `zlib-ng` of how to build a shared library. A better workaround than what I have now, is to have the `zlib-ng` package modify the `ldflags` by adding `-Wl,--undefined-version` when using `cce@17` -- older versions of `cce` may be affected too but I'm not sure when it started using LLD.

If you prefer this, I can replace the current workaround with the above -- what's the easiest way to modify the `ldflags` in `zlib-ng`?

Reply:
> what's the easiest way to modify the ldflags in zlib-ng?

To answer this directly: there is an existing `flag_handler` method - you can add a case for `ldflags`/`cce@17`:

```
    def flag_handler(self, name, flags):
        ...
        if name == "ldflags" and self.spec.satisfies("%cce@17"):
            flags.append("-Wl,--undefined-version")
        return (flags, None, None)
```

> uses the LLD linker which fails the autodetection of zlib-ng of how to build a shared library.

I'd be hopeful there might be some way to help it succeed at this. Then we could potentially address other cases of compilers using `lld`. That doesn't have to happen here, but the error message that comes up might be useful for future reference.

Reply:
The error reported by `zlib-ng`'s confugure script is:
```console
Checking for shared library support... No shared library support; try without defining CC and CFLAGS
```
The error message from the failed command in `configure.log` is:
```console
ld.lld: error: version script assignment of 'ZLIB_1.2.0' to symbol 'compressBound' failed: symbol not defined
```
The latter error message is repeated a number of times with different versions and symbols.

Adding the flag `-Wl,--undefined-version` is basically telling the linker to ignore this specific type of errors.

Another package that needed this flag is `util-linux-uuid` -- without it the build just failed with similar error message. At the moment, I'm just setting the `ldflags` for the compiler to include this flag for all packages. To get some of the `strumpack` executables to link (but probably not actually work), I also added `-Wl,--allow-shlib-undefined` to the compiler `ldflags`. For me this combination of `ldflags: -Wl,--undefined-version,--allow-shlib-undefined` was sufficient to resolve the LLD issues I encountered.

Reply:
Addressed in 1dbc0f61f6.

Reply:
> Another package that needed this flag is util-linux-uuid -- without it the build just failed with similar error message. 

I don't see that in https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/util-linux-uuid/package.py, or a recent PR on it. Was that in the works? Or did I miss something?

I think it might be worth following up with looking into how to avoid this, because I wonder if a `zlib-ng` library built without version definitions would potentially break other packages that would expect otherwise. Not something that has to be addressed here though.

Reply:
I haven't proposed a fix for `util-linux-uuid` for building with LLD or specifically with `cce` -- I just noticed in my test of this PR and resolved the issue locally by setting `ldflags: -Wl,--undefined-version,--allow-shlib-undefined` in the `cce@17.0.1` compiler definition I have in `etc/spack/compilers.yaml`.

If you want, I can add the fix for that from `zlib-ng` to `util-linux-uuid` as well.